ssl/tls13_enc.c

```diff
@@ -214,4 +213,228 @@ int tls13_generate_master_secret(SSL *s, unsigned char *out,
+    static const unsigned char server_handshake_traffic[] =
+        "server handshake traffic secret";
+    static const unsigned char server_application_traffic[] =
+        "server application traffic secret";
+    unsigned char key[EVP_MAX_KEY_LENGTH];
+    unsigned char iv[EVP_MAX_IV_LENGTH];
+    unsigned char secret[EVP_MAX_MD_SIZE];
+    unsigned char *insecret;
+    unsigned char *finsecret = NULL;
+    EVP_CIPHER_CTX *ciph_ctx;
+    const EVP_CIPHER *ciph = s->s3->tmp.new_sym_enc;;
+    size_t ivlen, keylen, finsecretlen;
+    const unsigned char *label;
+    size_t labellen;
+
+    if (which & SSL3_CC_READ) {
+        if (s->enc_read_ctx != NULL) {
+            EVP_CIPHER_CTX_reset(s->enc_read_ctx);
+        } else {
+            s->enc_read_ctx = EVP_CIPHER_CTX_new();
+            if (s->enc_read_ctx == NULL) {
+                SSLerr(SSL_F_TLS13_CHANGE_CIPHER_STATE, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+        }
+        ciph_ctx = s->enc_read_ctx;
+
+        RECORD_LAYER_reset_read_sequence(&s->rlayer);
+    } else {
+        if (s->enc_write_ctx != NULL) {
+            EVP_CIPHER_CTX_reset(s->enc_write_ctx);
+        } else {
+            s->enc_write_ctx = EVP_CIPHER_CTX_new();
+            if (s->enc_write_ctx == NULL) {
+                SSLerr(SSL_F_TLS13_CHANGE_CIPHER_STATE, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+        }
+        ciph_ctx = s->enc_write_ctx;
+
+        RECORD_LAYER_reset_write_sequence(&s->rlayer);
+    }
+
+    if (((which & SSL3_CC_CLIENT) && (which & SSL3_CC_WRITE))
+            || ((which & SSL3_CC_SERVER) && (which & SSL3_CC_READ))) {
+        if (which & SSL3_CC_HANDSHAKE) {
+            insecret = s->handshake_secret;
+            finsecret = s->client_finished_secret;
+            finsecretlen = sizeof(s->client_finished_secret);
+            label = client_handshake_traffic;
+            labellen = sizeof(client_handshake_traffic) - 1;
+        } else {
+            insecret = s->session->master_key;
+            label = client_application_traffic;
+            labellen = sizeof(client_application_traffic) - 1;
+        }
+    } else {
+        if (which & SSL3_CC_HANDSHAKE) {
+            insecret = s->handshake_secret;
+            finsecret = s->server_finished_secret;
+            finsecretlen = sizeof(s->server_finished_secret);
+            label = server_handshake_traffic;
+            labellen = sizeof(server_handshake_traffic) - 1;
+        } else {
+            insecret = s->session->master_key;
+            label = server_application_traffic;
+            labellen = sizeof(server_application_traffic) - 1;
+        }
+    }
+
+    if (!tls13_derive_secret(s, insecret, label, labellen, secret)) {
+        SSLerr(SSL_F_TLS13_CHANGE_CIPHER_STATE, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    /* TODO(size_t): convert me */
+    keylen = EVP_CIPHER_key_length(ciph);
+
+    if (EVP_CIPHER_mode(ciph) == EVP_CIPH_GCM_MODE)
```

Comment:
consider using a switch on the cipher mode, for future expansion?  not a requirement.